data/blog/2022-08-01---10-things.md

```diff
@@ -0,0 +1,144 @@
+
+I discovered that I could take the symmetric difference of two ItemsView objects, which gave me the items which were either a) new, or b) had changed. Precisely what I needed, but done at the C level. In Textual it is used to get the modified regions of the screen when a CSS property has changed, so that only the optimal regions need to be updated.
+
+For those who prefer to see the code, the following gist demonstrates the technique.
+
+<script src="https://gist.github.com/willmcgugan/61993ecf362a298bebfae10f49dea906.js"></script>
+
+### lru_cache is fast
+
+Perhaps not surprising given that `lru_cache` is literally designed to speed up your code, but `@lru_cache` is _fast_. I was surprised how fast it was.
+
+If you aren't familiar with `lru_cache` it is a decorator which you will find in the `functools` module in the standard library. Add it to a method and it will cache the return value of a function. If you set the `maxsize` parameter it will ensure your cache doesn't grow indefinitely.
+
+I was looking in to [the implementation](https://github.com/python/cpython/blob/main/Lib/functools.py#L566) of lru_cache in the CPython repos and I figured I could beat it. Spoiler: I couldn't. It turns out there was a [C version](https://github.com/python/cpython/blob/main/Modules/_functoolsmodule.c#L992) which makes it very fast for both cache hits and cache misses.
+
+Knowing it is that fast convinced me to lower the barrier to use `@lru_cache`. There are a number of small functions in Textual, that are not exactly slow, but called a large number of times. Many of them were highly cacheable and judicious use of `@lru_cache` provided a significant win. Typically a `maxsize` of around 1000-4000 was enough to ensure that the majority calls were cache hits.
+
+Here's an example of the kind of function that benefited from caching. This method combines two rectangular regions in to a single region that fits both. You can see it doesn't do a great deal of work, but it was called 1000s of times.
+
+<script src="https://gist.github.com/willmcgugan/8d161a6b1104aa2b79b6656893d4055d.js"></script>
+
+A word of advice when using `lru_cache`: always check your assumptions by inspecting the `cache_info()`. For effective caching you should expect to see `hits` growing faster than `misses`.
+
+### Immutable is best
+
+Following on from the previous tip, I'd like to sing the praises of immutable objects. Python doesn't have true immutable objects, but you can get much of the benefits from tuples, NamedTuples, or frozen dataclasses.
+
+It _seems_ like an arbitrary limitation that you can't change an object, but it rarely is in practice. Computer scientists will point out that many languages only have immutable objects.
```

Comment:
I'm guessing there are some functional languages that only have immutable objects? I would imagine there's some form of mutability though. I know of many that are immutable-first.